<a href="https://colab.research.google.com/github/LucasGleysson/Senac_ML/blob/master/ML_Aula_10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Árvore de decisão

- Análise de dados para aumentar o nível de satisfação dos clientes de um ecommerce utilizando árvore de decisão.

##Importação Pandas e Numpy

In [49]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

##Leitura das tabelas .csv

In [50]:
tabela_cliente = pd.read_csv('/content/drive/MyDrive/olist_customers_dataset.csv')
tabela_localizacao = pd.read_csv('/content/drive/MyDrive/olist_geolocation_dataset.csv')
tabela_pedido = pd.read_csv('/content/drive/MyDrive/olist_order_items_dataset.csv')
tabela_pagamento = pd.read_csv('/content/drive/MyDrive/olist_order_payments_dataset.csv')
tabela_review = pd.read_csv('/content/drive/MyDrive/olist_order_reviews_dataset.csv')
tabela_entrega_pedido = pd.read_csv('/content/drive/MyDrive/olist_orders_dataset.csv')
tabela_descricao_produto = pd.read_csv('/content/drive/MyDrive/olist_products_dataset.csv')
tabela_vendedor = pd.read_csv('/content/drive/MyDrive/olist_sellers_dataset.csv')
tabela_categoria_traduzido = pd.read_csv('/content/drive/MyDrive/product_category_name_translation.csv')

##Checagem .head()

###Cliente

In [51]:
tabela_cliente.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

###Localização

In [52]:
tabela_localizacao.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

###Pedido

In [53]:
tabela_pedido.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

###Pagamento

In [54]:
tabela_pagamento.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

###Review

In [55]:
tabela_review.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

###Entrega Pedido

In [56]:
tabela_entrega_pedido.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

###Descrição Produto

In [57]:
tabela_descricao_produto.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

###Vendedor

In [58]:
tabela_vendedor.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

###Categoria Traduzido

In [59]:
tabela_categoria_traduzido.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

##União das tabelas

- Inicio do processo de união das 9 tabelas disponibilizadas com a finalidade de produzir uma tabela resultante que possua os elementos mais importantes para a determinação do review_score. No primeiro merge realizado, unimos a tabela de clientes com as respectivas entregas dos pedidos usando o código individual de cada consumidor como parâmetro.

In [60]:
pd.merge(tabela_cliente, tabela_entrega_pedido, on=['customer_id'],how='inner')

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  \
0                         14409                 franca             SP   
1                          9790  sao bernardo do campo             SP   
2                          1151              sao paulo             SP   
3                          8775        mogi das cruzes             SP   
4                         13056               campinas             SP   
...                         ...                    ...            ...   
99436                      3937              sao paulo             SP   
99437                      6764        taboao da serra             SP   
99438                     60115              fortaleza             CE   
99439                     92120                 canoas             RS   
99440                      6703                  cotia             SP   

                               order_id order_status order_purchase_timestamp  \
0      00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1      29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2      b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3      951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   
...                                 ...          ...                      ...   
99436  6760e20addcf0121e9d58f2f1ff14298    delivered      2018-04-07 15:48:17   
99437  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered      2018-04-04 08:20:22   
99438  fed4434add09a6f332ea398efd656a5c    delivered      2018-04-08 20:11:50   
99439  e31ec91cea1ecf97797787471f98a8c2    delivered      2017-11-03 21:08:33   
99440  28db69209a75e59f20ccbb5c36a20b90    delivered      2017-12-19 14:27:23   

         order_approved_at order_delivered_carrier_date  \
0      2017-05-16 15:22:12          2017-05-23 10:47:57   
1      2018-01-12 20:58:32          2018-01-15 17:14:59   
2      2018-05-20 16:19:10          2018-06-11 14:31:00   
3      2018-03-13 17:29:19          2018-03-27 23:22:42   
4      2018-07-29 10:10:09          2018-07-30 15:16:00   
...                    ...                          ...   
99436  2018-04-07 16:08:45          2018-04-11 02:08:36   
99437  2018-04-04 08:35:12          2018-04-05 18:42:35   
99438  2018-04-08 20:30:03          2018-04-09 17:52:17   
99439  2017-11-03 21:31:20          2017-11-06 18:24:41   
99440  2017-12-19 18:50:39          2017-12-21 19:17:21   

      order_delivered_customer_date order_estimated_delivery_date  
0               2017-05-25 10:35:35           2017-06-05 00:00:00  
1               2018-01-29 12:41:19           2018-02-06 00:00:00  
2               2018-06-14 17:58:51           2018-06-13 00:00:00  
3               2018-03-28 16:04:25           2018-04-10 00:00:00  
4               2018-08-09 20:55:48           2018-08-15 00:00:00  
...                             ...                           ...  
99436           2018-04-13 20:06:37           2018-04-25 00:00:00  


1- (Cliente, Entregas)

2- (1, Pedidos)

3- (2, Pagamentos)

4- (3, Review)

5- (4, Vendedor)

In [61]:
teste = pd.merge(tabela_cliente, tabela_entrega_pedido, on=['customer_id'], how='inner')
teste = pd.merge(teste, tabela_pedido, on=['order_id'], how='inner')
teste = pd.merge(teste, tabela_pagamento, on=['order_id'], how='inner')
teste = pd.merge(teste, tabela_review, on=['order_id'], how='inner')
teste = pd.merge(teste, tabela_vendedor, on=['seller_id'], how='inner')
teste

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
2       b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
3       4c06b42fbf7b97ab10779cda5549cd1c  07d190f123147d9e89d4b922543d7948   
4       8276de07ef25225d412b8462d73f8664  332cf4e83e16004ba7dca932ce82475b   
...                                  ...                               ...   
117324  eda10d7eb51a675d000956fa1aabc7dd  5c22dbb9d8b5385953f2f1e6d9d115a7   
117325  f0f671d4034e98cdf20f0c452d6db02b  6532316a8c1a7c7a569482db073530bb   
117326  b90b857a277b080d8078332b3636ab58  7f5f43a9394964ac51bc5d1e0204ead4   
117327  1ed0c832c2dd99570a59260e71768bdf  82d46759af0369aad49084bacf85a6c3   
117328  1ed0c832c2dd99570a59260e71768bdf  82d46759af0369aad49084bacf85a6c3   

        customer_zip_code_prefix    customer_city customer_state  \
0                          14409           franca             SP   
1                           1151        sao paulo             SP   
2                           8775  mogi das cruzes             SP   
3                          65075         sao luis             MA   
4                          90010     porto alegre             RS   
...                          ...              ...            ...   
117324                     39920         jordania             MG   
117325                      7231        guarulhos             SP   
117326                     76400           uruacu             GO   
117327                     37610      bom repouso             MG   
117328                     37610      bom repouso             MG   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
2       951670f92359f4fe4a63112aa7306eba    delivered   
3       997fd4b37386d10b57d4ac8cdec011a6    delivered   
4       b63fff1ef8b6201df68bb611ee41129e    delivered   
...                                  ...          ...   
117324  b494f624c631105b24fe497277085909    delivered   
117325  f73b31435ce6dec43df056154c39a1ce      shipped   
117326  d76fe905f4588ad69ab3b36527bb267d    delivered   
117327  51c6d2f460589fa7b65f2da51e860206    delivered   
117328  51c6d2f460589fa7b65f2da51e860206    delivered   

       order_purchase_timestamp    order_approved_at  \
0           2017-05-16 15:05:35  2017-05-16 15:22:12   
1           2018-05-19 16:07:45  2018-05-20 16:19:10   
2           2018-03-13 16:06:38  2018-03-13 17:29:19   
3           2018-07-08 09:59:19  2018-07-08 10:30:15   
4           2017-04-02 22:15:27  2017-04-04 05:55:19   
...                         ...                  ...   
117324      2018-08-01 18:58:49  2018-08-02 18:50:15   
117325      2018-08-20 12:37:54  2018-08-20 15:35:42   
117326      2018-08-18 20:59:58  2018-08-18 21:10:13   
117327      2017-11-14 12:04:09  2017-11-14 12:15:25   
117328      2017-11-14 12:04:09  2017-11-14 12:15:25   

       order_delivered_carrier_date  ... payment_value  \
0               2017-05-23 10:47:57  ...        146.87   
1               2018-06-11 14:31:00  ...        157.73   
2               2018-03-27 23:22:42  ...        173.30   
3               2018-07-25 15:12:00  ...        315.75   
4               2017-04-13 16:28:06  ...        146.21   
...                             ...  ...           ...   
117324          2018-08-03 15:39:00  ...         72.18   
117325          2018-08-24 16:48:00  ...        775.98   
117326          2018-08-20 16:49:00  ...        224.44   
117327          2017-11-27 20:44:47  ...        167.79   
117328          2017-11-27 20:44:47  ...         32.13   

                               review_id  review_score review_comment_title  \
0       88b8b52d46df026a9d1ad2136a59b30b             4                  NaN   
1       5ad6695d76ee186dc473c42706984d87             5        

##Limpeza de Colunas

In [62]:
teste = teste.drop(columns=["customer_unique_id"],axis=1)
teste = teste.drop(columns=["customer_city"],axis=1)
teste = teste.drop(columns=["customer_state"],axis=1)
teste = teste.drop(columns=["order_status"],axis=1)
teste = teste.drop(columns=["order_purchase_timestamp"],axis=1)
teste = teste.drop(columns=["order_approved_at"],axis=1)
teste = teste.drop(columns=["order_delivered_carrier_date"],axis=1)
teste = teste.drop(columns=["order_delivered_customer_date"],axis=1)
teste = teste.drop(columns=["order_estimated_delivery_date"],axis=1)
teste = teste.drop(columns=["shipping_limit_date"],axis=1)
teste = teste.drop(columns=["review_creation_date"],axis=1)
teste = teste.drop(columns=["review_answer_timestamp"],axis=1)
teste = teste.drop(columns=["seller_city"],axis=1)
teste = teste.drop(columns=["seller_state"],axis=1)
teste = teste.drop(columns=["review_comment_title"],axis=1)
teste = teste.drop(columns=["review_comment_message"],axis=1)

- Tabela resultante após a remoção de atributos não prooritários para o nível de satisfação dos clientes.

In [63]:
display(teste.head())
teste.shape

customer_id  customer_zip_code_prefix  \
0  06b8999e2fba1a1fbc88172c00ba8bc7                     14409   
1  4e7b3e00288586ebd08712fdd0374a03                      1151   
2  b2b6027bc5c5109e529d4dc6358b12c3                      8775   
3  4c06b42fbf7b97ab10779cda5549cd1c                     65075   
4  8276de07ef25225d412b8462d73f8664                     90010   

                           order_id  order_item_id  \
0  00e7ee1b050b8499577073aeb2a297a1              1   
1  b2059ed67ce144a36e2aa97d2c9e9ad2              1   
2  951670f92359f4fe4a63112aa7306eba              1   
3  997fd4b37386d10b57d4ac8cdec011a6              1   
4  b63fff1ef8b6201df68bb611ee41129e              1   

                         product_id                         seller_id   price  \
0  a9516a079e37a9c9c36b9b78b10169e8  7c67e1448b00f6e969d365cea6b010ab  124.99   
1  bd07b66896d6f1494f5b86251848ced7  7c67e1448b00f6e969d365cea6b010ab  139.94   
2  a5647c44af977b148e0a3a4751a09e2e  7c67e1448b00f6e969d365cea6b010ab  149.94   
3  8ed094bfe076c568f6bb10feada3f75d  7c67e1448b00f6e969d365cea6b010ab  199.98   
4  aba86c093ccdbac75b09111d57e50004  7c67e1448b00f6e969d365cea6b010ab  129.60   

   freight_value  payment_sequential payment_type  payment_installments  \
0          21.88                   1  credit_card                     2   
1          17.79                   1  credit_card                     7   
2          23.36                   1  credit_card                     1   
3         115.77                   1  credit_card                     6   
4          16.61                   1       boleto                     1   

   payment_value                         review_id  review_score  \
0         146.87  88b8b52d46df026a9d1ad2136a59b30b             4   
1         157.73  5ad6695d76ee186dc473c42706984d87             5   
2         173.30  059a801bb31f6aab2266e672cab87bc5             5   
3         315.75  2518803238cabbf63564593e4a8fc52d             1   
4         146.21  84467f1766775ba4c14fc25b3ca67905             4   

   seller_zip_code_prefix  
0                    8577  
1                    8577  
2                    8577  
3                    8577  
4                    8577

(117329, 15)

##Otimização do DataFrame

In [64]:
vetor_cliente = np.array(teste.customer_id)
vetor_cepcliente = np.array(teste.customer_zip_code_prefix)
vetor_pedido = np.array(teste.order_id)
vetor_idpedido = np.array(teste.order_item_id)
vetor_produto = np.array(teste.product_id)
vetor_vendedor = np.array(teste.seller_id)
vetor_preco_produto = np.array(teste.price)
vetor_frete = np.array(teste.freight_value)
vetor_parcela = np.array(teste.payment_sequential)
vetor_tipopagamento = np.array(teste.payment_type)
vetor_pay = np.array(teste.payment_installments)
vetor_valorfinal = np.array(teste.payment_value)
vetor_review = np.array(teste.review_id)
vetor_score = np.array(teste.review_score)
vetor_cepvendedor = np.array(teste.seller_zip_code_prefix)

- DataFrame vazio

In [65]:
df = pd.DataFrame()
df

Empty DataFrame
Columns: []
Index: []

In [66]:
COLUNAS = [
'Cliente',
'CEP_Cliente',
'Pedido',
'id_Pedido',
'Produto',
'Vendedor',
'Preco_produto',
'Frete',
'Parcela',
'Tipo_pagamento',
'Installments',
'Valor_total',
'ID_Review',
'CEP_Vendedor',
'Score'
]

In [67]:
df = pd.DataFrame(columns =COLUNAS)
df.Cliente = vetor_cliente
df.CEP_Cliente = vetor_cepcliente
df.Pedido = vetor_pedido
df.id_Pedido = vetor_idpedido
df.Produto = vetor_produto
df.Vendedor = vetor_vendedor
df.Preco_produto = vetor_preco_produto
df.Frete = vetor_frete
df.Parcela = vetor_parcela
df.Tipo_pagamento = vetor_tipopagamento
df.Installments = vetor_pay
df.Valor_total = vetor_valorfinal
df.ID_Review = vetor_review
df.CEP_Vendedor = vetor_cepvendedor
df.Score = vetor_score
df

Cliente  CEP_Cliente  \
0       06b8999e2fba1a1fbc88172c00ba8bc7        14409   
1       4e7b3e00288586ebd08712fdd0374a03         1151   
2       b2b6027bc5c5109e529d4dc6358b12c3         8775   
3       4c06b42fbf7b97ab10779cda5549cd1c        65075   
4       8276de07ef25225d412b8462d73f8664        90010   
...                                  ...          ...   
117324  eda10d7eb51a675d000956fa1aabc7dd        39920   
117325  f0f671d4034e98cdf20f0c452d6db02b         7231   
117326  b90b857a277b080d8078332b3636ab58        76400   
117327  1ed0c832c2dd99570a59260e71768bdf        37610   
117328  1ed0c832c2dd99570a59260e71768bdf        37610   

                                  Pedido  id_Pedido  \
0       00e7ee1b050b8499577073aeb2a297a1          1   
1       b2059ed67ce144a36e2aa97d2c9e9ad2          1   
2       951670f92359f4fe4a63112aa7306eba          1   
3       997fd4b37386d10b57d4ac8cdec011a6          1   
4       b63fff1ef8b6201df68bb611ee41129e          1   
...                                  ...        ...   
117324  b494f624c631105b24fe497277085909          1   
117325  f73b31435ce6dec43df056154c39a1ce          1   
117326  d76fe905f4588ad69ab3b36527bb267d          1   
117327  51c6d2f460589fa7b65f2da51e860206          1   
117328  51c6d2f460589fa7b65f2da51e860206          1   

                                 Produto                          Vendedor  \
0       a9516a079e37a9c9c36b9b78b10169e8  7c67e1448b00f6e969d365cea6b010ab   
1       bd07b66896d6f1494f5b86251848ced7  7c67e1448b00f6e969d365cea6b010ab   
2       a5647c44af977b148e0a3a4751a09e2e  7c67e1448b00f6e969d365cea6b010ab   
3       8ed094bfe076c568f6bb10feada3f75d  7c67e1448b00f6e969d365cea6b010ab   
4       aba86c093ccdbac75b09111d57e50004  7c67e1448b00f6e969d365cea6b010ab   
...                                  ...                               ...   
117324  a2fce4213a6ac271129618259a334261  9140ef75cc8211c4035935e80e567204   
117325  5ece1e51ece4648de249ec6e9d007fd6  da4d149c0ddbac90557103ac0a0ec356   
117326  d6e7ac236e583af1e1543d8068c91676  377bc18334c78dd2f1535ef22a3ad225   
117327  c98bf47f7bea8f3aee82fa023786b8a1  51a04a8a6bdcb23deccc82b0b80742cf   
117328  c98bf47f7bea8f3aee82fa023786b8a1  51a04a8a6bdcb23deccc82b0b80742cf   

        Preco_produto   Frete  Parcela Tipo_pagamento  Installments  \
0              124.99   21.88        1    credit_card             2   
1              139.94   17.79        1    credit_card             7   
2              149.94   23.36        1    credit_card             1   
3              199.98  115.77        1    credit_card             6   
4              129.60   16.61        1         boleto             1   
...               ...     ...      ...            ...           ...   
117324          49.90   22.28        1    credit_card             3   
117325         730.00   45.98        1    credit_card            10   
117326         135.98   88.46        1    credit_card             6   
117327         167.99   31.93        1    credit_card             5   
117328         167.99   31.93        2    credit_card             1   

        Valor_total                         ID_Review  CEP_Vendedor  Score  
0            146.87  88b8b52d46df026a9d1ad2136a59b30b          8577      4  
1            157.73  5ad6695d76ee186dc473c42706984d87          8577      5  
2            173.30  059a801bb31f6aab2266e672cab87bc5          8577      5  
3            315.75  2518803238cabbf63564593e4a8fc52d          8577      1  
4            146.21  84467f1766775ba4c14fc25b3ca67905          8577      4  
...             ...                               ...           ...    ...  
117324        72.18  709ee020ff826ee5289d7dd5d6911d0c         87015      5  
117325       775.98  274e38de8990f74fc03894fa488b5cd5          9861      3  
117326       224.44  6010e1ba95a9e19dc10a7b761932cf98         83015      5  
117327       167.79  603f2873842a6975a43c54d305397d69         12914      1  
117328        32.13  603f2873842a6975a43c54d305397

- Isso pode otimizar as operações com o dataset por serem vetores de arrays do numpy agora.

In [68]:
df.Cliente

0         06b8999e2fba1a1fbc88172c00ba8bc7
1         4e7b3e00288586ebd08712fdd0374a03
2         b2b6027bc5c5109e529d4dc6358b12c3
3         4c06b42fbf7b97ab10779cda5549cd1c
4         8276de07ef25225d412b8462d73f8664
                        ...               
117324    eda10d7eb51a675d000956fa1aabc7dd
117325    f0f671d4034e98cdf20f0c452d6db02b
117326    b90b857a277b080d8078332b3636ab58
117327    1ed0c832c2dd99570a59260e71768bdf
117328    1ed0c832c2dd99570a59260e71768bdf
Name: Cliente, Length: 117329, dtype: object

##Informações do DataFrame

In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 117329 entries, 0 to 117328
Data columns (total 15 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Cliente         117329 non-null  object 
 1   CEP_Cliente     117329 non-null  int64  
 2   Pedido          117329 non-null  object 
 3   id_Pedido       117329 non-null  int64  
 4   Produto         117329 non-null  object 
 5   Vendedor        117329 non-null  object 
 6   Preco_produto   117329 non-null  float64
 7   Frete           117329 non-null  float64
 8   Parcela         117329 non-null  int64  
 9   Tipo_pagamento  117329 non-null  object 
 10  Installments    117329 non-null  int64  
 11  Valor_total     117329 non-null  float64
 12  ID_Review       117329 non-null  object 
 13  CEP_Vendedor    117329 non-null  int64  
 14  Score           117329 non-null  int64  
dtypes: float64(3), int64(6), object(6)
memory usage: 13.4+ MB


##Agrupando os elementos

In [70]:
df.groupby(by='Cliente').size()

Cliente
00012a2ce6f8dcda20d059ce98491703    1
000161a058600d5901f007fab4c27140    1
0001fd6190edaaf884bcaf3d49edf079    1
0002414f95344307404f0ace7a26f1d5    1
000379cdec625522490c315e70c7a9fb    1
                                   ..
fffcb937e9dd47a13f05ecb8290f4d3e    1
fffecc9f79fd8c764f843e9951b11341    3
fffeda5b6d849fbd39689bb92087f431    1
ffff42319e9b2d713724ae527742af25    1
ffffa3172527f765de70084a7e53aae8    2
Length: 97916, dtype: int64

##Import metodos de árvore de decisão

In [71]:
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.model_selection import train_test_split
from sklearn import metrics

- Para simplificar os dados e evitar criar um dummy com esse dataframe, removemos todos os elementos não-numéricos para que o modelo seja capaz de realizar a execução. Solução encotrada para simplificar os atributos do tipo "Objeto" para tipos numéricos.

In [72]:
df['Cliente'] = df['Cliente'].str.replace(r'\D','')
df['Pedido'] = df['Pedido'].str.replace(r'\D','')
df['Produto'] = df['Produto'].str.replace(r'\D','')
df['Vendedor'] = df['Vendedor'].str.replace(r'\D','')
df['ID_Review'] = df['ID_Review'].str.replace(r'\D','')
df

<ipython-input-72-25f4462c598d>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Cliente'] = df['Cliente'].str.replace(r'\D','')
<ipython-input-72-25f4462c598d>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Pedido'] = df['Pedido'].str.replace(r'\D','')
<ipython-input-72-25f4462c598d>:3: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Produto'] = df['Produto'].str.replace(r'\D','')
<ipython-input-72-25f4462c598d>:4: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Vendedor'] = df['Vendedor'].str.replace(r'\D','')
<ipython-input-72-25f4462c598d>:5: FutureWarning: The default value of regex will change from True to False in a future version.
  df['ID_Review'] = df['ID_Review'].str.replace(r'\D','')


Cliente  CEP_Cliente                      Pedido  \
0             068999211881720087        14409      0071050849957707322971   
1         4730028858608712037403         1151          205967144362972992   
2          260275510952946358123         8775      9516709235944631127306   
3             406427971077955491        65075         9974373861057480116   
4       827607252254128462738664        90010          631862016861141129   
...                          ...          ...                         ...   
117324          1075167500095617        39920  49462463110524497277085909   
117325       0671403498200452602         7231         7331435643056154391   
117326  908572770808078332363658        76400        76905458869336527267   
117327     108322995705926071768        37610      5162460589765251860206   
117328     108322995705926071768        37610      5162460589765251860206   

        id_Pedido                   Produto                   Vendedor  \
0               1    9516079379936978101698       76714480069693656010   
1               1    0766896614945862518487       76714480069693656010   
2               1     564744977148034751092       76714480069693656010   
3               1          8094076568610375       76714480069693656010   
4               1       8609375091115750004       76714480069693656010   
...           ...                       ...                        ...   
117324          1  242136271129618259334261  9140758211403593580567204   
117325          1         51514648249690076         414909055710300356   
117326          1    6723658311543806891676      377183347821535223225   
117327          1         98477838202378681           5104862382080742   
117328          1         98477838202378681           5104862382080742   

        Preco_produto   Frete  Parcela Tipo_pagamento  Installments  \
0              124.99   21.88        1    credit_card             2   
1              139.94   17.79        1    credit_card             7   
2              149.94   23.36        1    credit_card             1   
3              199.98  115.77        1    credit_card             6   
4              129.60   16.61        1         boleto             1   
...               ...     ...      ...            ...           ...   
117324          49.90   22.28        1    credit_card             3   
117325         730.00   45.98        1    credit_card            10   
117326         135.98   88.46        1    credit_card             6   
117327         167.99   31.93        1    credit_card             5   
117328         167.99   31.93        2    credit_card             1   

        Valor_total                   ID_Review  CEP_Vendedor  Score  
0            146.87        88852460269121365930          8577      4  
1            157.73     56695761864734270698487          8577      5  
2            173.30         0598013162266672875          8577      5  
3            315.75      2518803238635645934852          8577      1  
4            146.21     84467176677541425367905          8577      4  
...             ...                         ...           ...    ...  
117324        72.18        70902082652897569110         87015      5  
117325       775.98       274388990740389448855          9861      3  
117326       224.44       601019591910776193298         83015      5  
117327       167.79  60328738426975435430539769         12914      1  
117328        32.13  60328738426975435430539769         12914      1  

[117329 rows x 15 columns]

- Realizamos o procedimento de remoção dos elementos não-numéricos para todos as colunas do tipo objeto com exceção do tipo de pagamento pois o tipo de pagamento se resume a poucas opções. Dessa forma, usamos a função get_dummies apenas para o tipo de pagamento.
Portanto, a coluna Tipo_pagamento se divide em quatro colunas com lógica booleana. As novas colunas são: Tipo_pagamento_boleto, Tipo_pagamento_credit_card, Tipo_pagamento_debit_card, Tipo_pagamento_voucher.

In [73]:
result_df = pd.get_dummies(df, columns=['Tipo_pagamento'])
result_df

Cliente  CEP_Cliente                      Pedido  \
0             068999211881720087        14409      0071050849957707322971   
1         4730028858608712037403         1151          205967144362972992   
2          260275510952946358123         8775      9516709235944631127306   
3             406427971077955491        65075         9974373861057480116   
4       827607252254128462738664        90010          631862016861141129   
...                          ...          ...                         ...   
117324          1075167500095617        39920  49462463110524497277085909   
117325       0671403498200452602         7231         7331435643056154391   
117326  908572770808078332363658        76400        76905458869336527267   
117327     108322995705926071768        37610      5162460589765251860206   
117328     108322995705926071768        37610      5162460589765251860206   

        id_Pedido                   Produto                   Vendedor  \
0               1    9516079379936978101698       76714480069693656010   
1               1    0766896614945862518487       76714480069693656010   
2               1     564744977148034751092       76714480069693656010   
3               1          8094076568610375       76714480069693656010   
4               1       8609375091115750004       76714480069693656010   
...           ...                       ...                        ...   
117324          1  242136271129618259334261  9140758211403593580567204   
117325          1         51514648249690076         414909055710300356   
117326          1    6723658311543806891676      377183347821535223225   
117327          1         98477838202378681           5104862382080742   
117328          1         98477838202378681           5104862382080742   

        Preco_produto   Frete  Parcela  Installments  Valor_total  \
0              124.99   21.88        1             2       146.87   
1              139.94   17.79        1             7       157.73   
2              149.94   23.36        1             1       173.30   
3              199.98  115.77        1             6       315.75   
4              129.60   16.61        1             1       146.21   
...               ...     ...      ...           ...          ...   
117324          49.90   22.28        1             3        72.18   
117325         730.00   45.98        1            10       775.98   
117326         135.98   88.46        1             6       224.44   
117327         167.99   31.93        1             5       167.79   
117328         167.99   31.93        2             1        32.13   

                         ID_Review  CEP_Vendedor  Score  \
0             88852460269121365930          8577      4   
1          56695761864734270698487          8577      5   
2              0598013162266672875          8577      5   
3           2518803238635645934852          8577      1   
4          84467176677541425367905          8577      4   
...                            ...           ...    ...   
117324        70902082652897569110         87015      5   
117325       274388990740389448855          9861      3   
117326       601019591910776193298         83015      5   
117327  60328738426975435430539769         12914      1   
117328  60328738426975435430539769         12914      1   

        Tipo_pagamento_boleto  Tipo_pagamento_credit_card  \
0                           0                           1   
1                           0                           1   
2                           0                           1   
3                           0                           1   
4                           1                           0   
...                       ...                         ...   
117324                      0                           1   
117325                      0                           1   
117326                      0                           1   
117327                      0                           1   
11

##Criação de reserva para possíveis Conclusões

In [74]:
reserva = result_df.copy()
reserva

Cliente  CEP_Cliente                      Pedido  \
0             068999211881720087        14409      0071050849957707322971   
1         4730028858608712037403         1151          205967144362972992   
2          260275510952946358123         8775      9516709235944631127306   
3             406427971077955491        65075         9974373861057480116   
4       827607252254128462738664        90010          631862016861141129   
...                          ...          ...                         ...   
117324          1075167500095617        39920  49462463110524497277085909   
117325       0671403498200452602         7231         7331435643056154391   
117326  908572770808078332363658        76400        76905458869336527267   
117327     108322995705926071768        37610      5162460589765251860206   
117328     108322995705926071768        37610      5162460589765251860206   

        id_Pedido                   Produto                   Vendedor  \
0               1    9516079379936978101698       76714480069693656010   
1               1    0766896614945862518487       76714480069693656010   
2               1     564744977148034751092       76714480069693656010   
3               1          8094076568610375       76714480069693656010   
4               1       8609375091115750004       76714480069693656010   
...           ...                       ...                        ...   
117324          1  242136271129618259334261  9140758211403593580567204   
117325          1         51514648249690076         414909055710300356   
117326          1    6723658311543806891676      377183347821535223225   
117327          1         98477838202378681           5104862382080742   
117328          1         98477838202378681           5104862382080742   

        Preco_produto   Frete  Parcela  Installments  Valor_total  \
0              124.99   21.88        1             2       146.87   
1              139.94   17.79        1             7       157.73   
2              149.94   23.36        1             1       173.30   
3              199.98  115.77        1             6       315.75   
4              129.60   16.61        1             1       146.21   
...               ...     ...      ...           ...          ...   
117324          49.90   22.28        1             3        72.18   
117325         730.00   45.98        1            10       775.98   
117326         135.98   88.46        1             6       224.44   
117327         167.99   31.93        1             5       167.79   
117328         167.99   31.93        2             1        32.13   

                         ID_Review  CEP_Vendedor  Score  \
0             88852460269121365930          8577      4   
1          56695761864734270698487          8577      5   
2              0598013162266672875          8577      5   
3           2518803238635645934852          8577      1   
4          84467176677541425367905          8577      4   
...                            ...           ...    ...   
117324        70902082652897569110         87015      5   
117325       274388990740389448855          9861      3   
117326       601019591910776193298         83015      5   
117327  60328738426975435430539769         12914      1   
117328  60328738426975435430539769         12914      1   

        Tipo_pagamento_boleto  Tipo_pagamento_credit_card  \
0                           0                           1   
1                           0                           1   
2                           0                           1   
3                           0                           1   
4                           1                           0   
...                       ...                         ...   
117324                      0                           1   
117325                      0                           1   
117326                      0                           1   
117327                      0                           1   
11

- Eliminando todas as linhas com nível 4 ou nível 5 de satisfação. Dessa forma, temos um dataframe com todos os dados, e um apenas com
dados classificados com nível 3,2 ou 1 ou seja, mediano, ruim ou péssimo. (elementos que apresentam nível de insatisfação interessante
para análise).

In [75]:
reserva = reserva.drop(reserva[reserva.Score > 3].index)
reserva

Cliente  CEP_Cliente                   Pedido  \
3             406427971077955491        65075      9974373861057480116   
7         0372810041857927016263        21010    369525670169482825831   
18             45800571938861445        13485      3793048208056190003   
19          19530467435951219042        20920  54813422788333035171945   
21             52814168460899591        36091       532771277634465595   
...                          ...          ...                      ...   
117320  049671902463228846651580         4929     93358729701080188455   
117323     211018165721598633401        18650          357035504504584   
117325       0671403498200452602         7231      7331435643056154391   
117327     108322995705926071768        37610   5162460589765251860206   
117328     108322995705926071768        37610   5162460589765251860206   

        id_Pedido                   Produto                 Vendedor  \
3               1          8094076568610375     76714480069693656010   
7               1    9149964056487642250089     76714480069693656010   
18              1  073816586646007781918834     76714480069693656010   
19              1     466668028836324039256     76714480069693656010   
21              1            44685813716436     76714480069693656010   
...           ...                       ...                      ...   
117320          1      72966155721853046438    929923957665903146465   
117323          1   44442901095795215892510  46565251517718193618677   
117325          1         51514648249690076       414909055710300356   
117327          1         98477838202378681         5104862382080742   
117328          1         98477838202378681         5104862382080742   

        Preco_produto   Frete  Parcela  Installments  Valor_total  \
3              199.98  115.77        1             6       315.75   
7               78.98   39.48        1             1       118.46   
18             209.99   36.98        1             4       246.97   
19             159.98   49.49        1             1       209.47   
21             129.94   73.86        1             4       203.80   
...               ...     ...      ...           ...          ...   
117320         109.00   12.67        1             1       121.67   
117323         226.60   25.67        1             5       252.27   
117325         730.00   45.98        1            10       775.98   
117327         167.99   31.93        1             5       167.79   
117328         167.99   31.93        2             1        32.13   

                         ID_Review  CEP_Vendedor  Score  \
3           2518803238635645934852          8577      1   
7          50240342442053299029167          8577      2   
18         19338660916893707370094          8577      3   
19           576586473748925439048          8577      1   
21            47648253909483401757          8577      1   
...                            ...           ...    ...   
117320         4396354052945949758          4310      1   
117323          220719962586900027          9895      3   
117325       274388990740389448855          9861      3   
117327  60328738426975435430539769         12914      1   
117328  60328738426975435430539769         12914      1   

        Tipo_pagamento_boleto  Tipo_pagamento_credit_card  \
3                           0                           1   
7                           1                           0   
18                          0                           1   
19                          1                           0   
21                          0                           1   
...                       ...                         ...   
117320                      1                           0   
117323                      0                           1   
117325                      0                           1   
117327                      0                           1   
117328                      0                           1   



##Processo de separação de treino/teste

###Proporções estabelecidas
**70% Treino**

**30% Teste**

In [76]:
x_train, x_teste, y_train, y_teste = train_test_split(result_df.drop('Score', axis=1),result_df['Score'], test_size=0.3)

##Número de amostras para cada processo

In [77]:
x_train.shape, x_teste.shape

((82130, 17), (35199, 17))

##Número de targets para cada processo

In [78]:
y_train.shape, y_teste.shape

((82130,), (35199,))

##Criação do classificador

In [79]:
cls = DecisionTreeClassifier()

##Treinamento

In [80]:
cls = cls.fit(x_train, y_train)

##Vetor com as importancias de cada atributo para a determinação do review_score

In [81]:
cls.feature_importances_

array([0.11158023, 0.10974554, 0.1098609 , 0.00704168, 0.09005858,
       0.07403306, 0.07833494, 0.08455886, 0.00233602, 0.03923232,
       0.08806626, 0.11285103, 0.07125941, 0.00791166, 0.00768172,
       0.00228411, 0.0031637 ])

In [82]:
from sklearn.tree import DecisionTreeClassifier,plot_tree,export_graphviz,export_text

In [83]:
plt.rcParams['figure.figsize'] = (40,20)

In [84]:
'''plt.figure()
plot_tree(cls, filled=False)
plt.show()'''

'plt.figure()\nplot_tree(cls, filled=False)\nplt.show()'

##Trabalho de redução e melhoramento da visualização: notebook extra no final

In [85]:
df.head()

Cliente  CEP_Cliente                  Pedido  id_Pedido  \
0        068999211881720087        14409  0071050849957707322971          1   
1    4730028858608712037403         1151      205967144362972992          1   
2     260275510952946358123         8775  9516709235944631127306          1   
3        406427971077955491        65075     9974373861057480116          1   
4  827607252254128462738664        90010      631862016861141129          1   

                  Produto              Vendedor  Preco_produto   Frete  \
0  9516079379936978101698  76714480069693656010         124.99   21.88   
1  0766896614945862518487  76714480069693656010         139.94   17.79   
2   564744977148034751092  76714480069693656010         149.94   23.36   
3        8094076568610375  76714480069693656010         199.98  115.77   
4     8609375091115750004  76714480069693656010         129.60   16.61   

   Parcela Tipo_pagamento  Installments  Valor_total                ID_Review  \
0        1    credit_card             2       146.87     88852460269121365930   
1        1    credit_card             7       157.73  56695761864734270698487   
2        1    credit_card             1       173.30      0598013162266672875   
3        1    credit_card             6       315.75   2518803238635645934852   
4        1         boleto             1       146.21  84467176677541425367905   

   CEP_Vendedor  Score  
0          8577      4  
1          8577      5  
2          8577      5  
3          8577      1  
4          8577      4

**Para tornar o modelo mais visual, criamos um laço para a impressão dos pesoss de cada atributo para determinar o score**

In [86]:
for feature, importancia in zip(result_df.columns, cls.feature_importances_):
  print('{}: {:.1f}%'.format(feature,(importancia*100)))

Cliente: 11.2%
CEP_Cliente: 11.0%
Pedido: 11.0%
id_Pedido: 0.7%
Produto: 9.0%
Vendedor: 7.4%
Preco_produto: 7.8%
Frete: 8.5%
Parcela: 0.2%
Installments: 3.9%
Valor_total: 8.8%
ID_Review: 11.3%
CEP_Vendedor: 7.1%
Score: 0.8%
Tipo_pagamento_boleto: 0.8%
Tipo_pagamento_credit_card: 0.2%
Tipo_pagamento_debit_card: 0.3%


##Vetor com predições para chegatem do aprendizado

In [87]:
result = cls.predict(x_teste)
result

array([5, 4, 4, ..., 5, 4, 5])

##Representação das métricas de precisão e médias do modelo

In [88]:
from sklearn import metrics
print(metrics.classification_report(y_teste,result))

              precision    recall  f1-score   support

           1       0.26      0.26      0.26      4472
           2       0.13      0.13      0.13      1226
           3       0.15      0.16      0.16      3009
           4       0.24      0.26      0.25      6595
           5       0.61      0.59      0.60     19897

    accuracy                           0.43     35199
   macro avg       0.28      0.28      0.28     35199
weighted avg       0.44      0.43      0.44     35199



##Precisão Total

In [89]:
from sklearn.model_selection import cross_val_score
allScores = cross_val_score(cls, x_train, y_train, cv=10)
allScores.mean()

0.43141361256544497

**Treinamento utilizando o dataframe reserva.**

**(Apenas com os níveis de satisfação abaixo **
da média, score <=3)**

##Split do dataframe em treino e teste
**(70%, 30% respectiviamente)**

In [95]:
x_train, x_test, y_train, y_test = train_test_split(reserva.drop('Score', axis=1), reserva['Score'], test_size=0.3)

##Quantidade de amostrias do treino

In [96]:
x_train.shape, x_teste.shape

((20145, 17), (35199, 17))

##Classificador

In [97]:
clf = DecisionTreeClassifier()

##Treino

In [98]:
clf = clf.fit(x_train,y_train)


##Importancia de cada atributo para determinar o nível de satisfação dos consumidores

In [99]:
clf.feature_importances_

array([0.11012489, 0.11197204, 0.1051897 , 0.00448235, 0.08172171,
       0.08293559, 0.08380526, 0.0844557 , 0.00044774, 0.02966881,
       0.10096119, 0.1091792 , 0.07677175, 0.00656519, 0.00633777,
       0.00171289, 0.00366823])

In [104]:
for feature, importancia in zip(reserva.columns, clf.feature_importances_):
  print('{}: {:.1f}%'.format(feature,((importancia*100))))

Cliente: 11.0%
CEP_Cliente: 11.2%
Pedido: 10.5%
id_Pedido: 0.4%
Produto: 8.2%
Vendedor: 8.3%
Preco_produto: 8.4%
Frete: 8.4%
Parcela: 0.0%
Installments: 3.0%
Valor_total: 10.1%
ID_Review: 10.9%
CEP_Vendedor: 7.7%
Score: 0.7%
Tipo_pagamento_boleto: 0.6%
Tipo_pagamento_credit_card: 0.2%
Tipo_pagamento_debit_card: 0.4%


#Análise dos dados processados

- Com o objetivo de aumentar o review_score desse marketplace, utilizamos o algoritmo de árvore de decisão para encontar elementos que influenciam diretamente a aceitação e satisfação dos cliente, analisando desde o processo de despacho do produto até a qualidade do atendimento e produto final. Dessa forma, utilizamos estrtégias de filtragem. Inicialmente eliminamos os elementos que evidentemente, não apresentavam influiencia sobre a nota registrada pelo cliente. Após esse passo inciial, simplificamos o tipo dos dados para facilitar o processamento do dataframe pelo algoritmo. Após o processo de preparar os dados para o aprendizado restramos os seguintes valores:

Considerando todas as notas(ótimo, bom, mediano, ruim, péssimo):


Cliente: 11.0%

CEP_Cliente: 11.2%

Pedido: 10.5%

id_Pedido: 0.4%

Produto: 8.2%

Vendedor: 8.3%

Preco_produto: 8.4%

Frete: 8.4%

Parcela: 0.0%

Installments: 3.0%

Valor_total: 10.1%

ID_Review: 10.9%

CEP_Vendedor: 7.7%

Score: 0.7%

Tipo_pagamento_boleto: 0.6%

Tipo_pagamento_credit_card: 0.2%

Tipo_pagamento_debit_card: 0.4%


Portanto, com base nos resultados obtidos, o dono do marketplace deve se atentar aos seguintes parâmetros de sua logística:



1- Relação entre CEP_Cliente, Frete e CEP_Vendedor

Esses atributos apresentaram importancia direta na nota registrada pelo cliente, dessa forma, algumas razões devem ser consideradas:

- Problemas com a entrega
- Qualidade da entrega (tempo de entrega, comprometimento do produto durante o processo de transporte)
- Alto preço do frete para determinada regiões

2- Produto

O atributo "produto" apresentou importancia direta na insatisfação dos clientes. Portanto, deve-se considerar:

- Qualidade ruim de determinados produtos.
- O produto entrege não apresenta as características do produto anunciado.
-Entrega de produtos errados, problema de logística.

3- Vendedor
O atributo "vendedor" apresentou importancioa direta na insatisfação dos clientes. Portante, deve-se considerar:

- Quanlidade duvidosa do atendimento por parte do vendedor.
-Erro do vendedor em algumas etapas do processo em específico.

#Conclusão

Portanto, é possível concluir que para o marketplace aumentar seu score, analisar problemas logísticos por pare do processo de trnasporte, aspectos do produto e os vendedores são fatores mais importantestes para entender o que se passa na empresa e solucionar o problema que está gerando insatisfação dos cliuentes desse e-commerce.